# Limpieza de datos

* Vamos a aplicar el [checklist del Banco Mundial](https://dimewiki.worldbank.org/wiki/Checklist:_Data_Cleaning) en Python usando el [SciPy stack](https://www.scipy.org/stackspec.html), principalmente pandas

## 0. Prerequisitos

In [1]:
# Para que la notebook corra tanto en jupyter como en colab.
# NO ES RECOMENDADO realizar pip install desde un notebook.

import sys
in_colab = 'google.colab' in sys.modules

if in_colab:
  BASE_DIR = "https://github.com/DiploDatos/AnalisisYCuracion/raw/master/"
else:
  BASE_DIR = ".."
if 'ftfy' not in sys.modules:
    !pip install 'ftfy<5.6'

El sistema no puede encontrar el archivo especificado.


# 1. Importando los datos

# 1.1. Verificar que no hay problemas en la importación

In [2]:
import pandas as pd

# para que los numeros flotantes me los muestre solo con dos decimales
pd.options.display.float_format = '{:.2f}'.format

In [3]:
# Importamos datos de proyectos de Kickstarter la plataforma de Crowdsourcing

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv")

In [4]:
kickstarter_2018[:5]

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2  1000004038                                     Where is Hank?   
3  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4  1000011046  Community Film Project: The Art of Neighborhoo...   

         category main_category currency    deadline     goal  \
0          Poetry    Publishing      GBP  2015-10-09  1000.00   
1  Narrative Film  Film & Video      USD  2017-11-01 30000.00   
2  Narrative Film  Film & Video      USD  2013-02-26 45000.00   
3           Music         Music      USD  2012-04-16  5000.00   
4    Film & Video  Film & Video      USD  2015-08-29 19500.00   

              launched  pledged     state  backers country  usd pledged  \
0  2015-08-11 12:12:28     0.00    failed        0      GB         0.00   
1  2017-09-02 04:43:57  2421.00    failed       15      US       100.00   
2  2013-01-12 00:20:50   220.00    failed        3      US       220.00   
3  2012-03-17 03:24:11     1.00    failed        1      US         1.00   
4  2015-07-04 08:35:03  1283.00  canceled       14      US      1283.00   

   usd_pledged_real  usd_goal_real  
0              0.00        1533.95  
1           2421.00       30000.00  
2            220.00       45000.00  
3              1.00        5000.00  
4           1283.00       19500.00

In [5]:
# Set seed for reproducibility

# Al fijar la semilla no habrá azar.
# Esto es algo necesario cuando queremos "reproducir valores aleatorios".

import numpy as np
np.random.seed(0)

In [6]:
kickstarter_2018.sample(5)

ID                                               name  \
338862   796196901                                 10G Christmas Tree   
277871   483825010                                              Gliff   
47000    123916947                                 STUFFED Food Truck   
111338  1565733636  NeoExodus Adventure: Origin of Man for Pathfin...   
53743   1273544891        NAPOLEON IN NEW YORK! an original TV Series   

               category main_category currency    deadline     goal  \
338862              Art           Art      USD  2010-12-26 10526.00   
277871  Gaming Hardware         Games      USD  2016-03-28 10000.00   
47000       Food Trucks          Food      USD  2015-01-06 60000.00   
111338   Tabletop Games         Games      USD  2012-05-01   500.00   
53743            Comedy  Film & Video      USD  2016-07-26 25000.00   

                   launched  pledged       state  backers country  \
338862  2010-12-08 08:44:04     0.00      failed        0      US   
277871  2016-01-28 04:56:18    51.00      failed        5      US   
47000   2014-11-07 02:24:36    25.00      failed        1      US   
111338  2012-03-15 01:16:10   585.00  successful       17      US   
53743   2016-05-27 00:07:25    25.00      failed        1      US   

        usd pledged  usd_pledged_real  usd_goal_real  
338862         0.00              0.00       10526.00  
277871        51.00             51.00       10000.00  
47000         25.00             25.00       60000.00  
111338       585.00            585.00         500.00  
53743         25.00             25.00       25000.00

* Veamos si la descripción del dataset se corresponde con lo levantado    https://www.kaggle.com/kemical/kickstarter-projects/data

In [7]:
pd.DataFrame([["ID", "No description provided", "Numeric"],
["name", "No description provided", "String"],
["category", "No description provided", "String"],
["main_category", "No description provided", "String"],
["currency", "No description provided", "String"],
["deadline", "No description provided", "DateTime"],
["goal", "Goal amount in project currency", "Numeric"],
["launched", "No description provided", "DateTime"],
["pledged", "Pledged amount in the project currency", "Numeric"],
["state", "No description provided", "String"],
["backers", "No description provided", "Numeric"],
["country", "No description provided", "String"],
["usd pledged", "Pledged amount in USD (conversion made by KS)", "Numeric"],
["usd_pledged_real", "Pledged amount in USD (conversion made by fixer.io api)", "Numeric"],
["usd_goal_real", "Goal amount in USD", "Numeric"]], columns=["Field name","Field description", "Type"])

Field name                                  Field description  \
0                 ID                            No description provided   
1               name                            No description provided   
2           category                            No description provided   
3      main_category                            No description provided   
4           currency                            No description provided   
5           deadline                            No description provided   
6               goal                    Goal amount in project currency   
7           launched                            No description provided   
8            pledged             Pledged amount in the project currency   
9              state                            No description provided   
10           backers                            No description provided   
11           country                            No description provided   
12       usd pledged      Pledged amount in USD (conversion made by KS)   
13  usd_pledged_real  Pledged amount in USD (conversion made by fixe...   
14     usd_goal_real                                 Goal amount in USD   

        Type  
0    Numeric  
1     String  
2     String  
3     String  
4     String  
5   DateTime  
6    Numeric  
7   DateTime  
8    Numeric  
9     String  
10   Numeric  
11    String  
12   Numeric  
13   Numeric  
14   Numeric

* Ahora veamos los tipos de datos que detectó pandas

In [8]:
kickstarter_2018.dtypes

ID                    int64
name                 object
category             object
main_category        object
currency             object
deadline             object
goal                float64
launched             object
pledged             float64
state                object
backers               int64
country              object
usd pledged         float64
usd_pledged_real    float64
usd_goal_real       float64
dtype: object

* Los campos object generalmente son String
* Vemos que no reconoció como fechas a las variables: **deadline** y **launched** 

In [9]:
# Vemos un resumen de los datos

kickstarter_2018.describe()

ID         goal     pledged   backers  usd pledged  \
count     378661.00    378661.00   378661.00 378661.00    374864.00   
mean  1074731191.99     49080.79     9682.98    105.62      7036.73   
std    619086204.32   1183391.26    95636.01    907.19     78639.75   
min         5971.00         0.01        0.00      0.00         0.00   
25%    538263516.00      2000.00       30.00      2.00        16.98   
50%   1075275634.00      5200.00      620.00     12.00       394.72   
75%   1610148624.00     16000.00     4076.00     56.00      3034.09   
max   2147476221.00 100000000.00 20338986.27 219382.00  20338986.27   

       usd_pledged_real  usd_goal_real  
count         378661.00      378661.00  
mean            9058.92       45454.40  
std            90973.34     1152950.06  
min                0.00           0.01  
25%               31.00        2000.00  
50%              624.33        5500.00  
75%             4050.00       15500.00  
max         20338986.27   166361390.71

In [10]:
# Por defecto se ven los datos numéricos
# Pediremos ver el resto de los datos

kickstarter_2018.describe(include=['object'])

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   

          deadline             launched   state country  
count       378661               378661  378661  378661  
unique        3164               378089       6      23  
top     2014-08-08  1970-01-01 01:00:00  failed      US  
freq           705                    7  197719  292627

In [11]:
# Operamos sobre los datos pidiendo el valor mínimo

kickstarter_2018['launched'].min()

'1970-01-01 01:00:00'

In [12]:
# Operamos sobre los datos pidiendo el valor máximo
kickstarter_2018['launched'].max()

'2018-01-02 15:02:31'

In [13]:
# Indicamos que columnas corresponden a fechas

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"])
kickstarter_2018.dtypes

ID                           int64
name                        object
category                    object
main_category               object
currency                    object
deadline            datetime64[ns]
goal                       float64
launched            datetime64[ns]
pledged                    float64
state                       object
backers                      int64
country                     object
usd pledged                float64
usd_pledged_real           float64
usd_goal_real              float64
dtype: object

In [14]:
# Ahora podemos calcular el rango de la fecha de lanzamiento -> 'launched'

kickstarter_2018['launched'].max() - kickstarter_2018['launched'].min()

Timedelta('17533 days 14:02:31')

# 1.2. Asegurar de tener ids/claves únicas

* Chequeo de datos duplicados

In [15]:
kickstarter_2018.shape

(378661, 15)

* Pandas soporta índices (id) en los DataFrames

In [16]:
# Recargamos el conjunto de datos indicando que columna será el índice (index_col=['ID'])

kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"],
                               index_col=['ID'])

In [17]:
kickstarter_2018[:5]

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  
ID                         
1000002330        1533.95  
1000003930       30000.00  
1000004038       45000.00  
1000007540        5000.00  
1000011046       19500.00

In [18]:
kickstarter_2018.shape

(378661, 14)

* Indicar el índice nos permite que podamos usarlo para buscar por ese dato

In [19]:
kickstarter_2018.loc[999988282]

name                Nou Renmen Ayiti!  We Love Haiti!
category                              Performance Art
main_category                                     Art
currency                                          USD
deadline                          2011-08-16 00:00:00
goal                                          2000.00
launched                          2011-07-19 09:07:47
pledged                                        524.00
state                                          failed
backers                                            17
country                                            US
usd pledged                                    524.00
usd_pledged_real                               524.00
usd_goal_real                                 2000.00
Name: 999988282, dtype: object

* Buscar por índice, nos permite verificar si hay filas de contenidos duplicado

In [20]:
# Para buscar duplicados

kickstarter_2018[kickstarter_2018.duplicated()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

* No detectó duplicados
* Como Pandas acepta valores duplicados en los índices también debemos verificarlos

In [21]:
# Estadísticos descriptivos de la columna ID

pd.Series(kickstarter_2018.index,dtype=str).describe()

count       378661.00
mean    1074731191.99
std      619086204.32
min           5971.00
25%      538263516.00
50%     1075275634.00
75%     1610148624.00
max     2147476221.00
Name: ID, dtype: float64

In [22]:
# Buscamos nuevamente si hay datos duplicados

kickstarter_2018[kickstarter_2018.index.duplicated()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

* No detectó duplicados buscando por ID

## Repasamos con un ejemplo

In [23]:
csv='1,2\n3,3\n1,3'
print(csv)

1,2
3,3
1,3


In [24]:
# pasamos los datos a una tabla
from io import StringIO
df = pd.read_csv(StringIO(csv), names=['id','value'], index_col='id')
df

value
id       
1       2
3       3
1       3

In [25]:
# Buscamos duplicados
df[df.duplicated()]

value
id       
1       3

In [26]:
df[df.index.duplicated(keep=False)]

value
id       
1       2
1       3

### Ejercicio 1:

* Armar una tabla con todos los proyectos con nombres duplicados, ordenados para revisar agrupados. 

In [27]:
# Recargamos la base de datos indicando que la columna índice será el Nombre -> (index_col=['name'])

kickstarter_2018_name = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"],
                               index_col=['name'])
kickstarter_2018_name[:4]

ID        category  \
name                                                                            
The Songs of Adelaide & Abullah                    1000002330          Poetry   
Greeting From Earth: ZGAC Arts Capsule For ET      1000003930  Narrative Film   
Where is Hank?                                     1000004038  Narrative Film   
ToshiCapital Rekordz Needs Help to Complete Album  1000007540           Music   

                                                  main_category currency  \
name                                                                       
The Songs of Adelaide & Abullah                      Publishing      GBP   
Greeting From Earth: ZGAC Arts Capsule For ET      Film & Video      USD   
Where is Hank?                                     Film & Video      USD   
ToshiCapital Rekordz Needs Help to Complete Album         Music      USD   

                                                    deadline     goal  \
name                                                                    
The Songs of Adelaide & Abullah                   2015-10-09  1000.00   
Greeting From Earth: ZGAC Arts Capsule For ET     2017-11-01 30000.00   
Where is Hank?                                    2013-02-26 45000.00   
ToshiCapital Rekordz Needs Help to Complete Album 2012-04-16  5000.00   

                                                             launched  \
name                                                                    
The Songs of Adelaide & Abullah                   2015-08-11 12:12:28   
Greeting From Earth: ZGAC Arts Capsule For ET     2017-09-02 04:43:57   
Where is Hank?                                    2013-01-12 00:20:50   
ToshiCapital Rekordz Needs Help to Complete Album 2012-03-17 03:24:11   

                                                   pledged   state  backers  \
name                                                                          
The Songs of Adelaide & Abullah                       0.00  failed        0   
Greeting From Earth: ZGAC Arts Capsule For ET      2421.00  failed       15   
Where is Hank?                                      220.00  failed        3   
ToshiCapital Rekordz Needs Help to Complete Album     1.00  failed        1   

                                                  country  usd pledged  \
name                                                                     
The Songs of Adelaide & Abullah                        GB         0.00   
Greeting From Earth: ZGAC Arts Capsule For ET          US       100.00   
Where is Hank?                                         US       220.00   
ToshiCapital Rekordz Needs Help to Complete Album      US         1.00   

                                                   usd_pledged_real  \
name                                                                  
The Songs of Adelaide & Abullah                                0.00   
Greeting From Earth: ZGAC Arts Capsule For ET               2421.00   
Where is Hank?                                               220.00   
ToshiCapital Rekordz Needs Help to Complete Album              1.00   

                                                   usd_goal_real  
name                                                              
The Songs of Adelaide & Abullah                          1533.95  
Greeting From Earth: ZGAC Arts Capsule For ET           30000.00  
Where is Hank?                                          45000.00  
ToshiCapital Rekordz Needs Help to Complete Album        5000.00

In [28]:
# Creamos un nuevo dataset que tenga los registros con nombres duplicados
# Buscamos nuevamente si hay datos duplicados

kickstarter_2018_dup = kickstarter_2018[kickstarter_2018.name.duplicated(keep=False)].sort_values('name')
kickstarter_2018_dup[:4]

name     category  \
ID                                                                           
816998285   "...The Last shall be first..." LODB Lifestyle...      Fashion   
815783250   "...The Last shall be first..." LODB Lifestyle...      Fashion   
1010584633                                    "A Fresh Start"       Shorts   
713417995                                     "A Fresh Start"  Documentary   

           main_category currency   deadline    goal            launched  \
ID                                                                         
816998285        Fashion      USD 2014-02-09 6500.00 2013-12-19 21:51:50   
815783250        Fashion      USD 2014-03-01 6500.00 2014-01-03 20:39:16   
1010584633  Film & Video      USD 2011-09-25 3000.00 2011-08-28 21:27:52   
713417995   Film & Video      USD 2013-03-27 5000.00 2013-01-26 03:51:47   

            pledged       state  backers country  usd pledged  \
ID                                                              
816998285      0.00    canceled        0      US         0.00   
815783250     37.00      failed        3      US        37.00   
1010584633  3000.00  successful       24      US      3000.00   
713417995   1417.00      failed       26      US      1417.00   

            usd_pledged_real  usd_goal_real  
ID                                           
816998285               0.00        6500.00  
815783250              37.00        6500.00  
1010584633           3000.00        3000.00  
713417995            1417.00        5000.00

In [29]:
kickstarter_2018_dup.shape

(5125, 14)

* Tenemos **5125** registros con nombres duplicados

# 1.3. Despersonalizar datos y guardarlos en un nuevo archivo

Hay muchas técnicas para despersonalizar datos.

Para ilustrar mostramos las ofrecidas por Google https://cloud.google.com/dlp/docs/transformations-reference:

* **Reemplazo**: Reemplaza cada valor de entrada con un valor determinado.
* **Ocultamiento**: Quita un valor y lo oculta.
* **Enmascaramiento con caracteres**: Enmascara una string por completo o parcialmente mediante el reemplazo de un número determinado de caracteres con un carácter fijo especificado..
* **Seudonimización mediante el reemplazo de un valor de entrada con un hash criptográfico**: Reemplaza valores de entrada con una string hexadecimal de 32 bytes mediante una clave de encriptación de datos.
* **Cambio de fechas**: Cambia las fechas por un número de días al azar, con la opción de ser coherente en el mismo contexto..
* **Seudonimización mediante el reemplazo con token de preservación de formato criptográfico**: Reemplaza un valor de entrada con un token, o valor sustituto, de la misma longitud mediante la encriptación de preservación de formato (FPE) con el modo de operación FFX. Esto permite que se use el resultado en sistemas con validación de formato o que necesitan aparecer como reales a pesar de que la información no se revela.
* **Valores de depósito con base en rangos de tamaño fijos**: Enmascara los valores de entrada y los reemplaza por depósitos, o rangos dentro de los cuales se encuentra el valor de entrada.
* **Valores de depósito con base en rangos de tamaño personalizados**: Valores de entrada de depósito con base en rangos configurables por el usuario y valores de reemplazo.
* **Extracción de datos de tiempo**: EExtrae o preserva una porción de los valores Date, Timestamp y TimeOfDay.

In [30]:
# Librería para hacer hashs -> códigos para encriptar nombres y despersonalizar los datos
from hashlib import md5

In [31]:
# Es necesario setear el encoding utf-8 andes de hashear

# Los floats no tienen atributo 'encode' por eso los debemos pasar a str

def hashit(val):
    if isinstance(val, float): 
        return str(val)
    return md5(val.encode('utf-8')).hexdigest()


kickstarter_2018['name'].apply(hashit)

ID
1000002330    a6828ae8a2eca25f0dd7035efc0af0a0
1000003930    81609b3bdc0b96f429672d69702f2524
1000004038    c12f5c3bace2f0213cdb2679a265dca0
1000007540    4dbdcf09c86bbf5683ec39bc57b77f81
1000011046    9c01404a2ef702811c2088ce139042ad
1000014025    3a26b565c8396d4867dda27816836e04
1000023410    1602dce9289c950176c41ce124944335
1000030581    1abb0a27e1a3ce5a79965e3022c6bddc
1000034518    3b3a87aebec0d9c1b3a4a5e4da7a88a8
100004195     d7c53ef574a108d857d52087bb685244
100004721     eba20b66506f511bfacfe843176f5e15
100005484     9df25d0c0e2f8ea9de303f3d4efaf3c8
1000055792    717f62eae71408f00fa49227da83feae
1000056157    6532f0fa53550fd516921a2d7a1121d6
1000057089    799a510f874b6d47c1fde736d22cae3e
1000064368    93ebbffc751f9cc8d31e0147b7b1c7e3
1000064918    c10592831d537538905aefc12d54b49d
1000068480    0f109f9875645b801843775bbeadb1f1
1000070642    2d3d4b4267af87948bf7ac6d3ca13b54
1000071625    26c10e7de5df6269e71df1bf49ff4da9
1000072011    459cec5845add6a7f8fb1498cfc0e618
1000081649

### Ejercicio 2:

* Verificar que los proyectos que tienen nombres duplicados también tienen el hash de nombre duplicado

In [32]:
kickstarter_2018_dup.shape

(5125, 14)

In [33]:
# Verificamos si estos registros también tienen el hash de nombre duplicado

kickstarter_2018_dup['name_hash'] = kickstarter_2018_dup['name'].apply(hashit)
kickstarter_2018_dup[:4]

name     category  \
ID                                                                           
816998285   "...The Last shall be first..." LODB Lifestyle...      Fashion   
815783250   "...The Last shall be first..." LODB Lifestyle...      Fashion   
1010584633                                    "A Fresh Start"       Shorts   
713417995                                     "A Fresh Start"  Documentary   

           main_category currency   deadline    goal            launched  \
ID                                                                         
816998285        Fashion      USD 2014-02-09 6500.00 2013-12-19 21:51:50   
815783250        Fashion      USD 2014-03-01 6500.00 2014-01-03 20:39:16   
1010584633  Film & Video      USD 2011-09-25 3000.00 2011-08-28 21:27:52   
713417995   Film & Video      USD 2013-03-27 5000.00 2013-01-26 03:51:47   

            pledged       state  backers country  usd pledged  \
ID                                                              
816998285      0.00    canceled        0      US         0.00   
815783250     37.00      failed        3      US        37.00   
1010584633  3000.00  successful       24      US      3000.00   
713417995   1417.00      failed       26      US      1417.00   

            usd_pledged_real  usd_goal_real                         name_hash  
ID                                                                             
816998285               0.00        6500.00  0c7a251ffe4c7834cbc4b04906952ff1  
815783250              37.00        6500.00  0c7a251ffe4c7834cbc4b04906952ff1  
1010584633           3000.00        3000.00  67554ab4203d95f2f2f05365f768206e  
713417995            1417.00        5000.00  67554ab4203d95f2f2f05365f768206e

In [34]:
kickstarter_2018_dup.shape

(5125, 15)

* Tenemos **5125** registros con hash duplicados
* Estos registros se corresponden con los que habíamos detectado para nombres duplicados

# 1.4. Nunca modificar los datos crudos u originales


In [45]:
#if not in_colab:
#  kickstarter_2018.to_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv")


# 2. Pasos necesarios


In [46]:
# Paso necesario si no estamos en colab

import pandas as pd

if not in_colab:
  kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801-for-pandas.csv",
  #kickstarter_2018 = pd.read_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv",
                        index_col='ID',
                        parse_dates=['deadline','launched'])

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x99 in position 11: invalid start byte

In [37]:
kickstarter_2018.describe(include='all')

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   
first                        NaN             NaN           NaN      NaN   
last                         NaN             NaN           NaN      NaN   
mean                         NaN             NaN           NaN      NaN   
std                          NaN             NaN           NaN      NaN   
min                          NaN             NaN           NaN      NaN   
25%                          NaN             NaN           NaN      NaN   
50%                          NaN             NaN           NaN      NaN   
75%                          NaN             NaN           NaN      NaN   
max                          NaN             NaN           NaN      NaN   

                   deadline         goal             launched     pledged  \
count                378661    378661.00               378661   378661.00   
unique                 3164          nan               378089         nan   
top     2014-08-08 00:00:00          nan  1970-01-01 01:00:00         nan   
freq                    705          nan                    7         nan   
first   2009-05-03 00:00:00          nan  1970-01-01 01:00:00         nan   
last    2018-03-03 00:00:00          nan  2018-01-02 15:02:31         nan   
mean                    NaN     49080.79                  NaN     9682.98   
std                     NaN   1183391.26                  NaN    95636.01   
min                     NaN         0.01                  NaN        0.00   
25%                     NaN      2000.00                  NaN       30.00   
50%                     NaN      5200.00                  NaN      620.00   
75%                     NaN     16000.00                  NaN     4076.00   
max                     NaN 100000000.00                  NaN 20338986.27   

         state   backers country  usd pledged  usd_pledged_real  usd_goal_real  
count   378661 378661.00  378661    374864.00         378661.00      378661.00  
unique       6       nan      23          nan               nan            nan  
top     failed       nan      US          nan               nan            nan  
freq    197719       nan  292627          nan               nan            nan  
first      NaN       nan     NaN          nan               nan            nan  
last       NaN       nan     NaN          nan               nan            nan  
mean       NaN    105.62     NaN      7036.73           9058.92       45454.40  
std        NaN    907.19     NaN     78639.75          90973.34     1152950.06  
min        NaN      0.00     NaN         0.00              0.00           0.01  
25%        NaN      2.00     NaN        16.98             31.00        2000.00  
50%        NaN     12.00     NaN       394.72            624.33        5500.00  
75%        NaN     56.00     NaN      3034.09           4050.00       15500.00  
max        NaN 219382.00     NaN  20338986.27       20338986.27   166361390.71

## 2.1. En las etiquetas de las variables, no se deben usar caracteres especiales.

In [38]:
# helpful character encoding module
import chardet

¿Por qué? Por que aun hay limitaciones para trabajar con estos caracteres.

¿Cúales son los caracteres "normales"? 

In [39]:
import string

string.ascii_letters + string.digits

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

¿Qué es ASCII? Es una de las primeras codificaciones (o encoding) disponibles. Ver https://en.wikipedia.org/wiki/ASCII

Un encoding es un mapa de caracteres a una representación en bits (por ejemplo 1000001). ASCII es uno de los primeros estandares para interoperatividad entre computadoras (antes cada fabricante de computadora usa su propia codificación), contempla sólo los caracteres ingleses y usa 7 bits, por ejemplo *1000001* codifica el caracter *A*

Cómo sólo representaban los caracteres ingleses, empiezan a aparecer variantes para distintos idiomas, Microsoft hizo los propios para internacionalizar Windows, hubo otro estandar ISO 8859, hasta que llegó el éstandar Unicode que hizo múltiples encodings pero cada uno contempla múltiples idiomas e incluso es extensible a futuro. UTF-8 es de esos encodings el más utilizado. Ver https://en.wikipedia.org/wiki/Unicode

En particular, Python 3 utiliza UTF-8 por defecto (lo que no sucedía en Python 2).

Veamos un ejemplo práctico

In [40]:
# start with a string
before = "This is the euro symbol: €"

# check to see what datatype it is
type(before)

str

In [41]:
# encode it to a different encoding, replacing characters that raise errors
after = before.encode("utf-8", errors = "replace")

# check the type
type(after)

bytes

In [42]:
# take a look at what the bytes look like
after

b'This is the euro symbol: \xe2\x82\xac'

In [43]:
# convert it back to utf-8
print(after.decode("utf-8"))

This is the euro symbol: €


In [44]:
# try to decode our bytes with the ascii encoding
print(after.decode("ascii"))

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 25: ordinal not in range(128)

In [ ]:
# start with a string
before = "This is the euro symbol: €"

# encode it to a different encoding, replacing characters that raise errors
after = before.encode("ascii", errors = "replace")

# convert it back to utf-8
print(after.decode("ascii"))

# We've lost the original underlying byte string! It's been 
# replaced with the underlying byte string for the unknown character :(

Este error ya lo vimos...

In [ ]:
kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv")

El error `UnicodeDecodeError` nos indica que el archivo no estaría en Unicode. Si no está en algún metadato o no fue indicado es dificil detectar el encoding, por suerte hay un paquete para esto en Python, que aplica unas heurísticas para detectar encoding.

In [ ]:
# In Colab we need to download data locally
if in_colab:
  !wget -c $BASE_DIR/input/kickstarter-projects/ks-projects-201612.csv
  DOWN_DIR = "./"
else:
  DOWN_DIR = BASE_DIR + "/input/kickstarter-projects/"
  
# look at the first ten thousand bytes to guess the character encoding
with open(DOWN_DIR + "/ks-projects-201612.csv", 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))

# check what the character encoding might be
result

Entonces le podemos indicar el encoding al leer el archivos

In [ ]:
# read in the file with the encoding detected by chardet
kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv", encoding='Windows-1252')

# look at the first few lines
kickstarter_2016.head()

Hay otro problema con los caracteres que es más sutíl: **[Mojibake](https://en.wikipedia.org/wiki/Mojibake)**

Sucede al leer UTF-8 como ASCII extendido (ie. `Windows-*`, `iso-8859-*` entre otros)

Ejemplo de mojibake:

In [ ]:
"cigüeña".encode('utf-8').decode('iso-8859-1')

Para esto vamos a usar un módulo específico [ftfy](https://ftfy.readthedocs.io/en/latest/). 

Que nos permite ordenar cadenas por rareza

In [ ]:
import ftfy.badness as bad

In [ ]:
def weird(val):
    if isinstance(val, float): 
        return 0
    return bad.sequence_weirdness(val)


kickstarter_2018['name_weirdness'] = kickstarter_2018['name'].apply(weird)
kickstarter_2018[kickstarter_2018['name_weirdness'] > 1]

Volviendo a la consigna original, chequeamos que no haya caracteres fuera de a-Z, 0-9 y _ en los nombres de columnas

In [ ]:
kickstarter_2018.columns[~kickstarter_2018.columns.str.match(r'^(\w+)$')]

In [ ]:
kickstarter_2018.usd_pledged.head()

In [ ]:
kickstarter_2018.columns = kickstarter_2018.columns.str.replace(' ', '_')
kickstarter_2018.head()

In [ ]:
kickstarter_2018.usd_pledged.head()

#### Ejercicio 3

Comparar la cantidad de nombres raros en kickstarter_2018 con la que obtenemos al cargar 'ks-projects-201801.csv' con encoding iso-8859-1.

## 2.2. Tratar valores faltantes

Veamos cuantos valores nulos tenemos

In [ ]:
# get the number of missing data points per column
missing_values_count = kickstarter_2018.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count[missing_values_count > 0]

Esta medida en sí, no dice nada más que con cuantos valores debemos lidiar. 

Primero tenemos que ver si en términos relativos su impacto.

In [ ]:
len(kickstarter_2018.dropna())/len(kickstarter_2018)

In [ ]:
len(kickstarter_2018.dropna(subset=['name']))/len(kickstarter_2018)

También debemos ver que significan en términos de representatividad e importancia de nuestro análisis.

In [ ]:
kickstarter_2018[kickstarter_2018.name.isnull()]

Ejemplo, cuantos de los proyectos suspendidos no tienen nombre

In [ ]:
len(kickstarter_2018[kickstarter_2018.state == 'suspended'].dropna(subset=['name'])) / \
    len(kickstarter_2018[kickstarter_2018.state == 'suspended'])

En este caso, creemos que es seguro descartar los proyectos sin nombre

In [ ]:
kickstarter_2018 = kickstarter_2018.dropna(subset=['name'])

In [ ]:
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

In [ ]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()]

In [ ]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()].describe(include='all')

Interesante, todos los datos pertenecen a un país indeterminado.

Todos los datos de ese "país" no tienen dinero prometido?

In [ ]:
kickstarter_2018[(kickstarter_2018.country == 'N,0"') & ~(kickstarter_2018.usd_pledged.isnull())].head()

A priori como los proyectos no tienen patrocinantes, completamos en 0 los valores

In [ ]:
kickstarter_2018 = kickstarter_2018.fillna(0)
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

In [ ]:
kickstarter_2018.shape

Más métodos disponibles en https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-missing y http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

## 2.3. Codificar variables

Para trabajar con los algoritmos de aprendizaje automático, las variables categóricas estas deben ser codificadas como variables numéricas, no como cadenas.

Para esta tarea también hay diferentes estrategias, dos comunes son: asociar cadena a número y asociar cadena a columna.

In [ ]:
from sklearn import preprocessing

In [ ]:
kickstarter_2018.describe(include=['O'])

In [ ]:
column='main_category'
# Create a label (category) encoder object
le = preprocessing.LabelEncoder()
# Fit the encoder to the pandas column
le.fit(kickstarter_2018[column])

In [ ]:
# View encoder mapping
dict(zip(le.classes_,le.transform(le.classes_)))

In [ ]:
# Apply the fitted encoder to the pandas column
kickstarter_2018[column] = le.transform(kickstarter_2018[column]) 
kickstarter_2018.head()

In [ ]:
# Reversing encoding
kickstarter_2018[column] = le.inverse_transform(kickstarter_2018[column]) 
kickstarter_2018.head()

In [ ]:
from sklearn.preprocessing import LabelBinarizer
other_column = 'state'

lb = LabelBinarizer()

lb_results = lb.fit_transform(kickstarter_2018[other_column])
pd.DataFrame(lb_results, columns=((other_column + '_') + pd.Series(lb.classes_))).head(10)

More about preprocessing in http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

#### Ejercicio 4

Codificar `currency` con ambas estrategias

## 2.4. No cambiar los nombres de las variables de la fuente de origen


## 2.5. Verificar la consistencia de las variables
Aplicar reglas de integridad


In [ ]:
kickstarter_2018.head()

¿Hay proyectos sin patrocinadores pero con plata prometida?

In [ ]:
kickstarter_2018[(kickstarter_2018.backers == 0) & (kickstarter_2018.usd_pledged > 0)]

¿Hay proyecto no exitosos que pasaron el objetivo?

In [ ]:
kickstarter_2018[(kickstarter_2018.state != 'successful') & (kickstarter_2018.pledged > kickstarter_2018.goal)]

¿Cómo es, en promedio, la correlación entre el objetivo y la plata obtenida?

In [ ]:
df = kickstarter_2018.groupby('state')[['goal','pledged']].aggregate('mean')
df['pledged'] / df['goal']

¿Cuánta plata se obtiene en promedio por estado y categoría?

In [ ]:
kickstarter_2018.pivot_table(values='usd_pledged',index='state',columns='main_category')

#### Ejercicio 5

1. ¿Hay proyecto éxitosos que no consiguieron el objetivo? Si hay, ¿Qué porcentaje sí y cuál no?
2. Calcular una tabla con la cantidad de proyectos por categoría principal y estado.

## 2.6. Identificar y documentar valores atípicos/outliers


Queremos analizar las características de los proyectos que obtienen dinero pero sin considerar los casos atípicos.

In [ ]:
%matplotlib inline
kickstarter_2018.usd_pledged.hist();

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged');

"There are many rules of thumb for how to define an outlier but there is no silver bullet. One rule of thumb is that any data point that is three standard deviations away from the mean of the same data point for all observations."

In [ ]:
outliers = kickstarter_2018[kickstarter_2018.usd_pledged > (kickstarter_2018.usd_pledged.mean() + 
                                                            3 * kickstarter_2018.usd_pledged.std())]
outliers

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged');

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged',by='state');

Analizar outliers por categorías

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged',by='state');
kickstarter_2018.boxplot(column='usd_goal_real',by='state');

#### Ejercicio 6

Calcular los valores atípicos de 'usd_goal_real' y graficar los boxplots, con y sin estos valores por categoría

## 2.7. Evaluar cómo comprimir los datos 
para su almacenamiento más eficiente.

En la actualidad, en las nubes el almacenamiento no suele ser problema sino la velocidad para leerlo que suele ser determinante a la hora de determinar el formato de almacenamiento.


## 2.8. Guardar el set de datos con un nombre informativo.


EL problema de la ingeniería de software